scripts/doc_downloader/downloader.py

```diff
@@ -0,0 +1,281 @@
+    return db_path
+
+def record_download(
+    db_path: str,
+    url: str,
+    local_path: str,
+    status: str = "success",
+    etag: Optional[str] = None,
+    last_modified: Optional[str] = None,
+):
+    """Record a download in the database."""
+    with sqlite3.connect(db_path) as conn:
+        cursor = conn.cursor()
+        cursor.execute(
+            "INSERT OR REPLACE INTO downloads (url, local_path, status, etag, last_modified, timestamp) VALUES (?, ?, ?, ?, ?, datetime('now'))",
+            (url, str(local_path), status, etag, last_modified),
+        )
+        conn.commit()
+
+def get_download_status(db_path: str, url: str) -> tuple:
+    """Get download status from database."""
+    with sqlite3.connect(db_path) as conn:
+        cursor = conn.cursor()
+        cursor.execute(
+            "SELECT etag, last_modified FROM downloads WHERE url = ? AND status = 'success'",
+            (url,),
+        )
+        result = cursor.fetchone()
+    return result or (None, None)
+
+# --- Network Functions ---
+
+async def fetch_page(session: aiohttp.ClientSession, url: str, semaphore: asyncio.Semaphore) -> Optional[str]:
+    """Fetch a single page."""
+    try:
+        async with semaphore:
+            async with session.get(url, timeout=30) as response:
+                if response.status == 200:
+                    return await response.text()
+                logger.warning("Failed to fetch %s: HTTP %s", url, response.status)
+                return None
+    except Exception as e:
+        logger.error("Error fetching %s: %s", url, e)
+        return None
+
+async def download_page(
+    session: aiohttp.ClientSession,
+    url: str,
+    output_dir: Path,
+    db_path: str,
+    semaphore: asyncio.Semaphore,
+    force: bool,
+    max_retries: int,
+    base_url: str,
+) -> tuple[str, bool]:
+    """Download a single page and save it."""
+    local_path = get_local_path(url, output_dir, base_url)
+    
+    for attempt in range(max_retries):
+        try:
+            async with semaphore:
+                async with session.get(url, timeout=30) as response:
+                    if response.status == 200:
+                        content = await response.text()
+                        with open(local_path, "w", encoding="utf-8") as f:
+                            f.write(content)
+                        record_download(db_path, url, str(local_path))
+                        logger.info("Downloaded %s -> %s", url, local_path)
+                        return url, True
+                    else:
+                        logger.warning("Failed to download %s: HTTP %s", url, response.status)
+                        if response.status == 404:
+                            break # Don't retry on 404
+        except Exception as e:
+            logger.error("Error downloading %s: %s", url, e)
+        
+        if attempt < max_retries - 1:
+            await asyncio.sleep(1)
```

Comment:
nit: exponential time sleep (but prob. not worth to implement it right now)

Reply:
`s/1/(2 ** attempt)/` should be fine, right?